sandbox/cgroup/kill.go

```diff
@@ -0,0 +1,97 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cgroup
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"syscall"
+)
+
+var KillSnapProcesses = killSnapProcessesImplV1
+
+func pickKillV1Impl() {
+	KillSnapProcesses = killSnapProcessesImplV1
+}
+
+func pickKillV2Impl() {
+	KillSnapProcesses = killSnapProcessesImplV2
+}
+
+var syscallKill = syscall.Kill
+
+func killProcessesInCgroup(dir string, signal syscall.Signal) error {
+	pids, err := pidsInFile(filepath.Join(dir, "cgroup.procs"))
+	if err != nil {
+		return err
+	}
+
+	var firstErr error
+	for _, pid := range pids {
+		err := syscallKill(pid, signal)
+		if err != nil && err != syscall.ESRCH && firstErr == nil {
+			firstErr = err
+		}
+	}
+
+	return firstErr
+}
+
+func killSnapProcessesImplV1(snapName string, signal syscall.Signal) error {
+	if err := freezeSnapProcessesImplV1(snapName); err != nil {
+		return err
+	}
+	defer thawSnapProcessesImplV1(snapName)
+
+	return killProcessesInCgroup(filepath.Join(freezerCgroupV1Dir, fmt.Sprintf("snap.%s", snapName)), signal)
+}
+
+func killSnapProcessesImplV2(snapName string, signal syscall.Signal) error {
+	action := func(dir string) error {
+		if signal == syscall.SIGKILL {
+			// Use cgroup.kill if it exists (requires linux 5.14+)
+			err := writeExistingFile(filepath.Join(dir, "cgroup.kill"), []byte("1"))
+			if err == nil || !os.IsNotExist(err) {
+				return err
+			}
+			// Fallback to classic freeze/kill/thaw if cgroup.kill doesn't exist
+		}
+		if err := freezeOneV2(dir); err != nil {
+			return err
+		}
+		defer thawOneV2(dir)
+		return killProcessesInCgroup(dir, signal)
+	}
+
+	var firstErr error
+	skipError := func(err error) bool {
+		if !os.IsNotExist(err) && firstErr == nil {
+			firstErr = err
+		}
+		return true
+	}
+
+	if err := applyToSnap(snapName, action, skipError); err != nil {
+		return err
+	}
+
+	return firstErr
```

Comment:
Could we use errors.Join now?

Reply:
master is still at go 1.18, errors.Join was introduced in 1.20